### ISSUE TITLE
(Feature request) Comfort option: fold side mirrors

### ISSUE THREAD

**Issue by AlexeyPribytkin** [tedsalmon/BlueBus#24]
Is it possible to add new comfort options for e38?
- fold side mirrors on car lock with lock button on the key.
- unfold side mirrors on engine start.

**Comment by 0x200h** [comment]
Second this - would be fantastic...

**Comment by sroth0616** [comment]
The only question I have is, does the E38 have the side mirror motor?

**Comment by AlexeyPribytkin** [comment]
Yes, it has. I usually fold mirrors with the dedicated button on mirror control panel.
![image](https://user-images.githubusercontent.com/2332036/88059708-e6aa0f80-cb6d-11ea-92da-30d331619c50.png)

**Comment by tedsalmon** [comment]
This is a planned feature :)

**Comment by Alpejohn** [comment]
This would be a great feature, but i would rather want it to "unfold" when you actually start the engine.
If it fold/unfold by lock/unlock it will fold/unfold everytime you open your car to, say get your wallet, glasses etc.
In my opinion that would be annoying. So have them unfold when you start the car, so one can get in and out without it doing its thing for no reason.

I think there is ibus messages that says "engine running" or something like that, but i cant say for sure. I guess it should be possible somehow tho. :-)

**Comment by tedsalmon** [comment]
@Alpejohn,

That's easy enough. The IKE broadcasts the ignition status via `0x11`, so I know when the engine is running :smiley: 

-Ted

**Comment by QSSaber** [comment]
What I am concerned is about the durability of the gears that move the mirrors.
Is it possible to fold mirrors via pressing the lock button for x amount of time?; like the comfort access feature.

I'm saying this because the E60/E63 M5/M6 does this. Even if you code the the fold mirrors option through NCSExpert, you don't get the mirrors to fold by simply locking the doors. But rather, you have to press on the lock button for a few consecutive seconds. 

Would this be possible?

Gabe

**Comment by Alpejohn** [comment]
> What I am concerned is about the durability of the gears that move the mirrors.
> Is it possible to fold mirrors via pressing the lock button for x amount of time?; like the comfort access feature.
> 
> I'm saying this because the E60/E63 M5/M6 does this. Even if you code the the fold mirrors option through NCSExpert, you don't get the mirrors to fold by simply locking the doors. But rather, you have to press on the lock button for a few consecutive seconds.
> 
> Would this be possible?
> 
> Gabe

Thats a good point, the gears in one of my mirrors are gone, and the other one is moving very slow or not at all.
So i need to overhaul them.
I guess it would be possible to activate the folding when the convenient closing activates, when you hold the locking button for a while and it closes the windows etc. But would be nice if it was possible to fold just by holding it for a few seconds, insteaad of the time it takes to activate the convenient closing.. But i dont know if it would be possible.

**Comment by 60SecTheBaptist** [comment]
I think the easiest solution is to provide the ability to Fold/Unfold with Lock/Unlock and menu options for just unfold if folded and disable the option.   for me i would like to have the fold unfold but i wouldn't use it until after i rebuild my motors.  otherwise im just looking for trouble. so option-disabled for me.

**Comment by harryberlin** [comment]
don't understand the discussion.
ted said it's easy.
the hold event should be easy, too.
so, just wait feature release and be happy.

**Comment by QSSaber** [comment]
> don't understand the discussion.
> ted said it's easy.
> the hold event should be easy, too.
> so, just wait for feature release and be happy.

Hi Harry.

Although folding mirrors are common features on the new cars, our old car's mirrors were not designed to fold and unfold everytime one locks or unlocks the door. 
If Ted said it would be easy to implement, that is great!
But this discussion is how to make the feature more fit for the old cars.
I think most of us are just throwing out ideas. Though I am not sure if this is meaningful to Ted.

**Comment by piersholt** [comment]
Excessive usage is a reasonable concern. The folding mirrors are driven by a team of particularly frail, asthmatic ants as it is, so ideally their demise is not accelerated.

> But would be nice if it was possible to fold just by holding it for a few seconds, insteaad of the time it takes to activate the convenient closing.. 

This is _technically_ possible, but likely unfeasible. The GM does broadcast what are effectively remote entry "press" and "release" events, but, any kind of timing based trigger would require a layer of logic. However, given the timing of the events is inconsistent, that can be quite tricky.

> I think a compromise might be to provide the ability to Fold/Unfold with Lock/Unlock and menu options for just unfold if folded and disable the option. for me i would like to have the fold unfold but i wouldn't use it until after i rebuild my motors. otherwise im just looking for trouble. so option-disabled for me.

The option to simply disable the feature is likely, even if only to limit unnecessary diagnostic traffic.

### .

All that being said, the feature mightn't be coming soon, or at least, not with support for all vehicles. ZKE encapsulates an... _absurd_, amount of modules, and there's multiple ZKE generations. Reverse engineering the diagnostics is what takes the time; the implementation itself is reasonably straight forward.

None of this is "easy" 😀

**Comment by vgergiev** [comment]
In case this really great option (not a sarcasm) will be added to the bluebus functionality, could you please consider to implement it as a separate option in settings and turned off by default?
Not sure how it is going in the US, but in Europe and CIS major part of the mirrors gears are destroyed (I've tried to replace them in 2 different services and without any luck).
Also, it would be great to have the following options for folding: Lock/Unlock, Engine Start/Stop, Disabled (default).

**Comment by Alpejohn** [comment]
> In case this really great option (not a sarcasm) will be added to the bluebus functionality, could you please consider to implement it as a separate option in settings and turned off by default?
> Not sure how it is going in the US, but in Europe and CIS major part of the mirrors gears are destroyed (I've tried to replace them in 2 different services and without any luck).
> Also, it would be great to have the following options for folding: Lock/Unlock, Engine Start/Stop, Disabled (default).

I know this isnt the right place to ask this, but since you mentioned it, what was the problem when you tried to swap the gears?
I have a set of new gears for both my mirrors. Already taken apart one to see how its done. Biggest thing i found was that the locking ring probably needs to be replaced. And i cleaned the bearing and filled it with greas. The mirror moved very easy, but i never tried  putting on a new locking ring tho.

If its possible, feel free to send me a PM if thats possible on ths site ( i dont know if it is but i would think so)**

Fredrik

**Comment by Chrisdavey1974** [comment]
Up until now I have been using my Intravee 2 and I love it! Originally I connected my iPod, but now I just use it for the features (power fold, auto lock, 3 blinks and coolant temp).  The last 5 years or so, I have been using the power fold feature on my E53. The best option for me was to fold on lock, and unfold on engine start. Used in this way, I’ve had no issues with motors. I also love the  way it can turn on lights when you unlock the car. I also love the PDC display, where it tells me in cm how far away I am. This helps combat dodgy sensors!   But there is no Bluetooth, hence BlueBus is the way forward for me. Best of both worlds (once power fold has been added) if as I do, you want to keep the stock look 👍

**Comment by StefPedro** [comment]
Will this feature see the light of day? It would be honestly really nice to not fiddle with the power folding mirror button all the time, and to not unlock and lock the car again because i forgot to fold the mirrors.

**Comment by DKeegan99** [comment]
Hi Ted, 

Checking on in this, as I'd love to have this configurable on my E39 M5. Happy to be a guinea pig, too, if needed. :)

- Doug
